src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,33 +122,143 @@ def __init__(
+            if not produce_billing.done():
+                break
+
+            if produce_billing.exception():
+                logger.error(
+                    "Async future failed in billing metrics consumer.",
+                    exc_info=produce_billing.exception(),
+                    extra={"offset": metrics_msg.offset},
+                )
+
+            self._ongoing_billing_outcomes.popleft()
+            self._ready_to_commit[metrics_msg.partition] = Position(
+                metrics_msg.next_offset, datetime.now()
+            )
+            self._messages_ready_since_last_commit += 1
+
+    def _should_commit(self) -> bool:
+        if not self._ready_to_commit:
+            return False
+        if self._messages_ready_since_last_commit >= self._max_batch_size:
+            return True
+        if self._last_commit_time + self._max_batch_time <= datetime.now():
+            return True
+        return False
+
+    def _bulk_commit(self) -> None:
+        """Commits and clears the ready to commit queue."""
+        if self._messages_ready_since_last_commit == 0:
+            return
+        self._commit(self._ready_to_commit)
+        self._clear_ready_queue()
+
+    def _clear_ready_queue(self) -> int:
+        """Clears the ready to commit queue and returns the amount of items dropped."""
+        amount_dropped = self._messages_ready_since_last_commit
+        self._ready_to_commit = {}
+        self._messages_ready_since_last_commit = 0
+        self._last_commit_time = datetime.now()
+        return amount_dropped
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+        self._bulk_commit()
+        while self._ongoing_billing_outcomes:
+            if deadline and deadline > time.time():
+                items_left = self._messages_ready_since_last_commit + len(
+                    self._ongoing_billing_outcomes
+                )
+                logger.warning(f"join timed out, items in the queue left: {items_left}")
+                break
+            self._mark_commit_ready()
+            self._bulk_commit()
+
     def terminate(self) -> None:
         self.close()
+        amount_dropped = len(self._ongoing_billing_outcomes)
+        while self._ongoing_billing_outcomes:
+            ongoing_work, _ = self._ongoing_billing_outcomes.popleft()
+            ongoing_work.cancel()
+        amount_dropped += self._clear_ready_queue()
+        if amount_dropped > 0:
+            logger.warning(f"terminated, items dropped: {amount_dropped}")
 
     def close(self) -> None:
-        self.__closed = True
+        self._closed = True
 
     def submit(self, message: Message[KafkaPayload]) -> None:
-        assert not self.__closed
-        self.__messages_since_last_commit += 1
+        if self._closed:
+            raise MessageRejected("Consumer is closed, no new messages are accepted.")
```

Comment:
From the documentation:

> If the processing strategy is unable to accept a message (due to it being at or over capacity, for example), this method will raise a `MessageRejected` exception.

I assume the consumer being closed also means it's unable to accept a message, so decided to raise the mentioned exception instead of asserting. It should be easier to debug.

Reply:
MessageRejected actually serves a different purpose. It tells the StreamProcessor to slow down if the processing strategy is busy and can't keep up.

If the consumer is closed then you should raise a different error as that is completely invalid. i.e. I think you can change this line to `assert not self._closed`.

Back to MessageRejected though. There is a scenario in which it should raise it. If the producer can't keep up then the list of futures may just keep growing faster than we are clearing it which can be dangerous. In that case you probably want to raise MessageRejected to slow down the consumer.

Since this (producing) is actually a fairly common behaviour for consumers, Arroyo's ProduceAndCommit step which is included out of the box already implements this here: https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/produce.py. You may want to consider just reusing this strategy instead of implementing your own. Note though that this will only be in the next release of Arroyo (which will probably today) and is not right now in the one that is used by Sentry.